src/components/flyout/flyout.tsx

```
@@ -347,10 +349,12 @@ export const EuiFlyout = forwardRef(
     }
 
     const isDefaultConfiguration = ownFocus && !isPushed;
-    const onClickOutside =
-      (isDefaultConfiguration && outsideClickCloses !== false) ||
+    const onClickOutside = (event: MouseEvent | TouchEvent) =>
+      (isDefaultConfiguration &&
+        outsideClickCloses !== false &&
+        event.target === maskRef.current) ||
       outsideClickCloses === true
-        ? onClose
+        ? onClose(event)
         : undefined;
```

This still feels pretty hard for me to read personally 😅 Since this is a function now, can we prefer early returns?
```tsx
const onClickOutside = (event: MouseEvent | TouchEvent) => {
  if (outsideClickCloses === true) return onClose(event);
  if (outsideClickCloses === false) return undefined;
  // If outsideClickCloses is not specified, only close on clicks to the overlay mask
  if (isDefaultConfiguration && event.target === maskRef) return onClose(event);
  // Otherwise if ownFocus is false, outside clicks should not close the flyout
  return undefined;
}

---

I'll refactor for readability, but here's the logic:

* If `ownFocus=true` (default), only clicks on the mask should close the flyout, even if `outsideClickCloses=true`
* If `ownFocus=false`, the flyout will only close if `outsideClickCloses=true`, and clicking any external element will close

---

> If ownFocus=true (default), only clicks on the mask should close the flyout, even if outsideClickCloses=true

That's just now what the current branching is doing though?? So you'd need to change this logic if that's the behavior you want. `outsideClickCloses === true` will always evaluate to returning `onClose()`. Am I reading this the same way as you?

---

```ts
const onClickOutside = (event: MouseEvent | TouchEvent) => {
  // Do not close the flyout for any external click
  if (outsideClickCloses === false) return undefined;
  if (isDefaultConfiguration) {
    // The overlay mask is present, so only clicks targeting the mask should close the flyout
    if (event.target === maskRef.current) return onClose(event);
  } else {
    // No overlay mask is present, so any outside clicks should close the flyout
    if (outsideClickCloses === true) return onClose(event);
  }
  // Otherwise if ownFocus is false, outside clicks should not close the flyout
  return undefined;
};
```

---

> That's just not what the current branching is doing though??

Yep I had the conditional wrong. https://github.com/elastic/eui/pull/5876#discussion_r867038126 is correct

---

@thompsongl Nice! Definitely much easier to grok now, especially w/ comments. For the last comment/return, I might make it a bit more specific to help follow the possible branch paths:

```
// Otherwise if ownFocus is false and outsideClickCloses is undefined, outside clicks should not close the flyout
```

---

I also think we should be more specific about the `isDefaultConfiguration` mask behavior:

```
    // The overlay mask is present, so only clicks on the mask should close the flyout, regardless of outsideClickCloses
```

---

Oh man, I remember how tricky this logic was. I'm going to repeat the logic just to ensure I understand.

- **Default** `ownFocus = true`: The overlay mask renders and the flyout will only be closed when clicking mask (or close button)
- `ownFocus = false && outsideClickCloses = false`: No overlay mask and can only be closed via close button (or some other in-page toggle)
- `ownFocus = false && outsideClickCloses = true`: Clicking anywhere outside will close flyout (does this include dragging with mouse up or does this now mean a special extra configuration?)

I _feel_ like that's how it was before, but maybe I'm remembering another config wrong?

---

> ownFocus = false && outsideClickCloses = true: Clicking anywhere outside will close flyout (does this include dragging with mouse up or does this now mean a special extra configuration?)

By default the close event will happen if mousedown happens outside the flyout. Consumers can change this to occur on mouseup if they have some reason to do so.
By default the close event will not happen if mousedown happens inside the flyout, even if mouse up happens outside the flyout (e.g., mousedown inside, drag outside, release. These flyout remains open)

---

[43ce709](https://github.com/elastic/eui/pull/5876/commits/43ce7096c703eea62af32303589f038d80018218) incorporates all suggestions in this thread